src/NodeHelper.groovy

```diff
@@ -0,0 +1,1136 @@
+/* (C) Copyright IBM Corporation 2018.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+import jenkins.model.Jenkins;
+import hudson.model.Computer;
+import hudson.util.RemotingDiagnostics;
+import hudson.remoting.Channel;
+import hudson.slaves.DumbSlave;
+import hudson.slaves.CommandLauncher;
+import hudson.model.Slave;
+import hudson.model.Node.Mode;
+import hudson.plugins.sshslaves.SSHLauncher;
+
+class  NodeHelper {
+
+    // TODO: Move strings out to a config file
+
+    /* Java Web Start (Windows)
+        - Internal data directory: remoting
+       SSH (most linux machines)
+        - Host: of the machine
+        - Credentials
+        - Host key verfication strategy: Non
+       Command launcer 
+        - only for machines that need to be reached via a proxy
+        - just one string with the machine ip
+     */
+
+    /**
+     * Adds a new node to jenkins.
+     *
+     * @param newNodeName         the name for the new node
+     * @param newNodeDescription  the description for the new node
+     * @param newNodeRemoteFS     the remote file structure
+     * @param newNodeNumExecutors the max number of executors allowed
+     * @param newNodeMode         the Hudson control constants which
+     *                            dictate how that control how much
+     *                            the node gets utilized. 
+     *                            Normal, use it as much as possible. 
+     *                            Exclusive, used it only for jobs that 
+     *                            specify this node as the assigned node
+     * @param newNodeLabelString  the labels to be added to the node. 
+     *                            Each label should be seperated by space
+     * @param launcher            the agent launcher for the node. Examples:
+     *                             - hudson.plugins.sshslaves.SSHLauncher
+     *                             - hudson.slaves.CommandLauncher
+     *
+     * @return name of the node that have just been created
+     *         <code>INVALID_NODE_NAME</code> if the node name
+     *         passed doesn't meet the required criteria.
+     */
+    public String addNewNode(
+        String newNodeName,
+        String newNodeDescription,
+        String newNodeRemoteFS,
+        int newNodeNumExecutors,
+        Mode newNodeMode,
+        String newNodeLabelString,
+        def launcher
+        ) {
+
+        String ret = "INVALID_NODE_NAME";
```

Comment:
ret --> nodeName?

Reply:
Same goes for all usages of 'ret', can replace with more meaningful variable name

Reply:
Using "ret" makes it easier to follow the return variable through a function. As for meaning, the function name, header comments, and function(s) called to update "ret" variable aid in the meaning.